src/librbd/DeepCopyRequest.cc

```diff
@@ -0,0 +1,383 @@
+      if (finish_op_ctx != nullptr) {
+        finish_op_ctx->complete(0);
+      }
+    });
+  librados::AioCompletion *comp = create_rados_callback(ctx);
+  int r = m_dst_image_ctx->md_ctx.aio_operate(m_dst_image_ctx->header_oid,
+                                              comp, &op);
+  assert(r == 0);
+  comp->release();
+  m_dst_image_ctx->snap_lock.put_read();
+  m_dst_image_ctx->owner_lock.put_read();
+}
+
+template <typename I>
+void DeepCopyRequest<I>::handle_set_size(uint64_t size, int r) {
+  ldout(m_cct, 20) << "r=" << r << dendl;
+
+  {
+    Mutex::Locker locker(m_lock);
+    if (r == 0 && m_canceled) {
+      r = -ECANCELED;
+    }
+  }
+
+  if (r < 0) {
+    if (r != -ECANCELED) {
+      lderr(m_cct) << "failed to copy snapshot metadata: " << cpp_strerror(r)
+                   << dendl;
+    }
+    finish(r);
+    return;
+  }
+
+  {
+    // adjust in-memory image size now that it's updated on disk
+    RWLock::WLocker snap_locker(m_dst_image_ctx->snap_lock);
+    m_dst_image_ctx->size = size;
+  }
+
+  send_copy_image();
+}
+
+template <typename I>
+void DeepCopyRequest<I>::send_copy_image() {
+  m_lock.Lock();
+  if (m_canceled) {
+    m_lock.Unlock();
+    finish(-ECANCELED);
+    return;
+  }
+
+  ldout(m_cct, 20) << dendl;
+
+  Context *ctx = create_context_callback<
+    DeepCopyRequest<I>, &DeepCopyRequest<I>::handle_copy_image>(this);
+  m_image_copy_request = ImageCopyRequest<I>::create(
+      m_src_image_ctx, m_dst_image_ctx, m_snap_id_start, m_snap_id_end,
+      m_object_number, *m_snap_seqs, m_prog_ctx, ctx);
+  m_image_copy_request->get();
+  m_lock.Unlock();
+
+  m_image_copy_request->send();
+}
+
+template <typename I>
+void DeepCopyRequest<I>::handle_copy_image(int r) {
+  ldout(m_cct, 20) << "r=" << r << dendl;
+
+  {
+    Mutex::Locker locker(m_lock);
+    m_image_copy_request->put();
+    m_image_copy_request = nullptr;
+    if (r == 0 && m_canceled) {
+      r = -ECANCELED;
+    }
+  }
+
+  if (r == -ECANCELED) {
+    ldout(m_cct, 10) << "image copy canceled" << dendl;
+    finish(r);
+    return;
+  } else if (r < 0) {
+    lderr(m_cct) << "failed to copy image: " << cpp_strerror(r) << dendl;
+    finish(r);
+    return;
+  }
+
+  send_copy_object_map();
+}
+
+template <typename I>
+void DeepCopyRequest<I>::send_copy_object_map() {
+  m_dst_image_ctx->owner_lock.get_read();
+  m_dst_image_ctx->snap_lock.get_read();
+  if (m_snap_id_end == CEPH_NOSNAP ||
+      !m_dst_image_ctx->test_features(RBD_FEATURE_OBJECT_MAP,
+                                      m_dst_image_ctx->snap_lock)) {
+    m_dst_image_ctx->snap_lock.put_read();
+    m_dst_image_ctx->owner_lock.put_read();
+    send_copy_metadata();
+    return;
+  }
+
+  assert(m_dst_image_ctx->object_map != nullptr);
+
+  ldout(m_cct, 20) << dendl;
+
+  Context *finish_op_ctx = nullptr;
+  if (m_dst_image_ctx->exclusive_lock != nullptr) {
+    finish_op_ctx = m_dst_image_ctx->exclusive_lock->start_op();
+  }
+  if (finish_op_ctx == nullptr) {
```

Comment:
If exclusive lock feature is enabled, this would be expected

Reply:
@dillaman I am not sure I see why it is expected? And actually it is the same we currently have in `rbd_mirror::ImageSync::send_copy_object_map()`:

[1] https://github.com/ceph/ceph/blob/master/src/tools/rbd_mirror/ImageSync.cc#L302

Reply:
... sorry, meant to say "is disabled, this would be expected". Unlike rbd-mirror which requires journaling and thus exclusive-lock, there shouldn't be such a restriction for deep-copy.

Reply:
But we are here only if object-map is enabled, which means exclusive-lock is enabled too?

Reply:
Indeed -- sorry for the noise.